### ISSUE TITLE
Removing dependency on a web browser for fly to access anything other than local authentication

### ISSUE THREAD

**Issue by FilBot3** [concourse/concourse#3208]
## What challenge are you facing?

The problem I am running into is being able to login using the `fly` application on a headless server to automate creation of Pipelines, or Teams without the use of a Web Browser.

## What would make this better?

Being able to use `fly` without having to open a web browser to authenticate against anything other than local authentication. Otherwise, being able to hit the HTTP API for authentication to use with a script to get the key, and then use fly for the rest of the commands it provides.

## Are you interested in implementing this yourself?

If I were proficient with Golang, I would.

**Comment by predatorian3** [comment]
Some existing conversation about this: https://discuss.concourse-ci.org/t/how-to-get-concourse-auth-token-when-i-login-concourse-in-the-without-browser-linux-server/242/11?u=chlorophil

**Comment by eedwards-sk** [comment]
more related conversations:

https://discuss.concourse-ci.org/t/automate-concourse-oauth-login/1023/3
https://discuss.concourse-ci.org/t/how-to-login-manually-in-concourse-4/1022/4

**Comment by eedwards-sk** [comment]
This is still a big issue for me. In fact I haven't yet even added a local user because I haven't figured out how I want to manage deployment and rotation of its credentials for automation.

The underlying issue I see is that I have to put credentials for a user with power to do everything in concourse, directly in the config on a web server.

**Comment by lrstanley** [comment]
Stalebot please.. still a large issue that needs to be addressed. Bot is driving me nuts!

**Comment by pivotal-jamie-klassen** [comment]
I think the underlying ask here is something along the lines of 'make it easier for my automation to interact with the Concourse API', so I'll link #2936 which I always do on such conversations

**Comment by eedwards-sk** [comment]
Reopen stalebot, you monster!

**Comment by eedwards-sk** [comment]
This is still a MASSIVE freaking issue. I do NOT want to have to specify credentials in a CONFIG FILE on a WEB SERVER!! D:

**Comment by krishnam-pakalapati-git** [comment]
We see this issue blocking us develop a plugin for Concourse or tools to automate Concourse jobs,pipeline management.

Any help on this would be highly appreciated.

**Comment by vito** [comment]
As often as this comes up, it's never really been made clear what needs to be changed or what feature needs to be added so that we end up with something that's actually an improvement over using local auth. :confused:

Usually the ask is to be able to log in without a browser and without local auth, but without saying *why* those two options are removed. It's presented as a puzzle, like "I'd like to be able to brush my teeth without using my hands." :smile: 

At some point you're going to be putting credentials on a server. Whether it's a generated token or a  local account with a generated password. So what needs to change about those credentials to make it better than local auth?

One problem with local auth which @eedwards-sk [pointed out](https://discuss.concourse-ci.org/t/automate-concourse-oauth-login/1023/3) is that local users have to be configured cluster-wide, which also requires a restart of Concourse. So let's get that one out of the way: is the idea that we can generate and revoke a credential through `fly` and provide that to a server instead?

What would the permissions for that credential be? Does that need to be configurable? Is that necessary for MVP? Does anyone want to share a proof-of-concept or anything so we can keep the ball rolling?

---

As a side note, the stale bot isn't meant to be the issue slayer, it's meant to keep the discussion developing into something that can actually come to be. I get that he's scary, and he doesn't communicate well, but - well, he's a bot. His feelings are synthetic and he's not great at empathy. But without the stale bot this project will drown in ideas that no one wants to solve, and it'll be really hard to actually prioritize things. I'd rather not go back to those days.

This idea clearly has a lot of people who care about it, but it needs to be more fleshed-out and understood before anything can be implemented. There isn't enough time in the day for myself and the team to go around to everyone's issues and flesh them out and develop them ourselves while also developing the Concourse roadmap. We need y'all to help when you can. :slightly_smiling_face:

When the stale bot comes around, I would genuinely appreciate it if y'all could heed its warning and try to push the idea further instead of complaining about the bot or leaving a 'bump' comment on the issue. Even if you don't have time to write a PR yourself, if this issue had a pretty clear idea to implement it, someone out there might find the issue who *does* have time, and then instead of waiting around for the core Concourse team to swoop in and pick it up the feature might actually be done quickly.

This project needs contributors - we have enough ideas for a lifetime!

That being said, I've added the 'help wanted' label to keep the bot at bay, since this request isn't going to go away, but before anything can be done we really need to figure out what exactly it is that the community needs here. Sorry that it took so many cycles of warning and closing before we came around and added the 'help wanted' label though - I've deleted the noisy comments.

Thanks for your patience y'all.

**Comment by lrstanley** [comment]
Going to be a long response, but figured I'd post here so I can reference in other places later...

> [...] it's meant to keep the discussion developing into something that can actually come to be. [...]

The response body doesn't really radiate this. The way it comes off in the response is"give us a PR or we won't bother".

> [...] without the stale bot this project will drown in ideas that no one wants to solve, and it'll be really hard to actually prioritize things. I'd rather not go back to those days.

Can't expect that every person coming along will have a contributor mindset. Even if the information is included or not (I think enough is included in this issue), there is still a large lack of response for many issues, which ruins the morale of people wanting to contribute more (I've seen this on many issues with heaps of information, which just end up getting closed with no response). This isn't a fault of the Concourse team, just an overall bandwidth issue, and ofc open source. Just an observation, nothing more.

Is there any way to adjust the bot to slow down a bit (in time), if there are more thumbs up style reactions to something? One common theme I see for things is people generally want a specific feature, but aren't going to be familiar with the open-source flow. I think it's fair to ask for more information, but I don't know if it's good to just close things with the lack of info (given clearly it's something many wish for).

----------------------------
As for the issue at hand...:

> As often as this comes up, it's never really been made clear what needs to be changed or what feature needs to be added so that we end up with something that's actually an improvement over using local auth. 😕

Automation:
> [...] being able to login using the fly application on a headless server **to automate creation of Pipelines, or Teams** without the use of a Web Browser.

> This is still a big issue for me. [...] **for automation**.

----------------------------
> Usually the ask is to be able to log in without a browser and without local auth, but without saying why those two options are removed.

Automation:
> because I haven't figured out how I want to manage deployment and rotation of its credentials for automation.

> I think the underlying ask here is something along the lines of **make it easier for my automation** to interact with the Concourse API, so I'll link #2936 which I always do on such conversations

> We see this issue blocking us develop a plugin for Concourse or tools **to automate Concourse jobs,pipeline management**. [...]

It's simple: today, the only way to do automation semi-reliably, is via the fly cli (there are advancements to job specific dynamic stuff, but this isn't really automation). Problem is, the current login flow is suited for users, not scripts/automation.

----------------------------
You might say that's what local users are for, but, as you pointed out, there are flaws with this approach:

> One problem with local auth which @eedwards-sk pointed out is that local users have to be configured cluster-wide, which also requires a restart of Concourse. So let's get that one out of the way: is the idea that we can generate and revoke a credential through fly and provide that to a server instead?

tl;dr: local user flow isn't great, and it needs to be improved in a way that doesn't require a bunch of disruption to the cluster. I.e, much in the same way we can manage things like LDAP permissions, we should be able to manage local users. Not cluster wide (this should still be supported for team management automation), but at a team level (teams can automate their own things, without granting higher level access).

----------------------------------
More so, an additional usecase is supporting all other flows of auth via the CLI. Even just supporting LDAP via the CLI would make sense. See no point in directing someone to a different window/tab/etc to input information, when they already have the window with the terminal selected, and are likely already about to type more things. Esp. if it's headless, fly is running from a different environment (e.g. vscode external service terminal, it's not running locally). Now, I still see a potential usecase for this, for those who are less savvy, however if they're already flying pipelines and understand the very complex nature of what it takes to push a pipeline to Concourse (steep learning curve for sure..), logging in the via the CLI seems like it'd be a piece of cake.

Our usecase today:
- We automate the create/destruction of teams via global automation, using an admin local user.
- We automate the monitoring of specific build steps/jobs/tasks/etc for compliance reasons, using an admin local user (have to login to each team individually, **over 150 teams**).
   - For each team, we use `get-pipeline` with `--json` on each pipeline, and run our scans (**2,700 pipelines** as of a week or so ago).
- For each team and unpaused pipeline, we also check build logs to see when the last run was. If it was a very long time ago, we pause the pipeline (reducing resource checks, given we have so many pipelines).
- We have product teams (product team maps to a concourse team, in this scenario), create/destroy pipelines for specific branches and whatnot (and the new things in 5.8.0 could potentially help with this), in addition to their own automation from external systems for new apps/apis/etc. We don't want to give these users **keys to the kingdom**.
- We have external systems (not managed by the team that manages Concourse) who create pipelines based off order forms/automated event systems, based off the team name (i.e. ability to give them access to create pipelines for specific teams without giving them access to all, e.g. one very sensitive team, is difficult). This type of team isn't a product team, more a group that manages pipelines for other teams.

Our normal developer flow for onboarding an app/api/etc, is:
1. PO (product owner) requests new team in our permission system.
2. Automation reads permission system and makes necessary changes to the environment (delta: adds team)
3. PO add developer permissions to a user for the necessary specific environments in permissions system.
4. Automation triggers delta change: add LDAP user to team (by setting all users at once).
5. Developer clones our template pipeline repo into their ci folder, which contains templates for things like:
   - Code/security scanning included.
   - Building of API's, apps, containers, and libraries and storing them in artifact store.
   - Deploying artifacts to various platforms.
   - Uses standard container images for runs.
   - Supports email (and other) alerts.
6. Developer runs ./ci/new.sh (or similar). This bootstraps by asking lots of questions and running a few tasks, like:
   1. What is your product/team name?
   2. What pipeline type is it?
   3. What environments will it deploy to?
   4. Pipeline name (if not already automatically determined)?
   5. Does it need a secret store? (see below)
   6. Prompts for user credentials, then uses the credentials in a loop to login to various systems (ensuring much easier audit trail), e.g:
      - Writes some bootstrap secrets to the secret store.
      - Makes some necessary VCS changes using https auth.
   7. Runs a string replacement on the template folders, moving files/folders around, auto injecting the team and various other configs. Also deleting the bootstrapping script.
8. Run `./login.sh` which will loop through each environment (dt, qa, prod), have each open a tab in their browser, type in their credentials three times, and often times (due to where fly is being ran)  copy/paste the token to the terminal three times.
9. Run `./update-pipeline.sh` which pushes the latest config to each environment as necessary/depending on specific prompts.

Ideally, 8 and 9 could have many of theirs steps completely removed, as we can just prompt for credentials and pass directly to the fly CLI.

Eventually, we'll likely move towards a Go-based CLI tool that will download the necessary pipeline templates from a branch and initializing, with support for upgrading, etc. This will help us determine who has customizations, what versions people are using, support more dynamic configurations, etc.

It also could be said that the environments we have could be merged, but I don't think that will end well, given our current scale, in addition to the other benefits to security, operation, upgrades, etc.

----------------------------------

tl;dr: things like forcing browser access, having undocumented API's, etc, seems (IMO) as though it's cutting off a good amount of the potential for Concourse, primarily in the enterprise space. I feel as though adoption would be much greater if there was a larger, more beginner-friendly ecosystem, as more companies move towards more automation. Not all Concourse users are going to be the stereotypical "using a macbook at a coffee shop, sipping my latte, writing my nodejs with emojis in each git commit message ⭐️🔥"

Rant over, sorry 😄

**Comment by vito** [comment]
@lrstanley Thanks for the details and no worries about the ranting, I'm in no position to judge there. :sweat_smile: Your comment really helps me understand the problem space more. :slightly_smiling_face: 

> The response body doesn't really radiate this. The way it comes off in the response is"give us a PR or we won't bother".

Yeah true, I'll adjust it a bit. We haven't really revised the text since we initially introduced it. Thanks for the feedback!

> Can't expect that every person coming along will have a contributor mindset. Even if the information is included or not (I think enough is included in this issue), there is still a large lack of response for many issues, which ruins the morale of people wanting to contribute more (I've seen this on many issues with heaps of information, which just end up getting closed with no response). This isn't a fault of the Concourse team, just an overall bandwidth issue, and ofc open source. Just an observation, nothing more.

I don't expect everyone to have a contributor mindset, which is exactly why I would like for issues like this one to at the very least develop *ideas* in the absence of contributors, without necessarily asking everyone to submit *PRs*. Over time, more and more people will see this issue. If the issue gets more detailed, the chance of one of those people actually being able to contribute increases.

If someone runs into the same need and finds a dead-end - whether it's a closed issue or an open issue with a bunch of +1's but no direction - they might just give up, even if they had the capability to contribute. If someone runs into the same need and finds an issue with a theoretical solution, they might try to implement it.

> Is there any way to adjust the bot to slow down a bit (in time), if there are more thumbs up style reactions to something? One common theme I see for things is people generally want a specific feature, but aren't going to be familiar with the open-source flow. I think it's fair to ask for more information, but I don't know if it's good to just close things with the lack of info (given clearly it's something many wish for).

Hmm it doesn't look like the bot has that kind of parameter. I'll look into adjusting the timing, but in my experience (and the [experience of other maintainers](https://github.com/probot/stale#is-closing-stale-issues-really-a-good-idea)) it's better to be explicit and close issues that aren't going anywhere. This is preferable to leaving it open and providing false hope of something maybe happening someday.

> Automation:

With respect, this much is obvious. If the use case was most accurately and completely described as "automation" I would be able to just tell you to use local users, because you can automate using those, but clearly this problem runs deeper than that. That's what I'm trying to get at. :slightly_smiling_face: 

For example, let's go over your use cases one-by-one:

> * We automate the create/destruction of teams via global automation, using an admin local user.

This sounds very interesting on its own, because we've been musing about pushing some of this automation into Concourse itself, with some early signs of that in the `set_pipeline` step - and later, perhaps something like ["projects"](https://blog.concourse-ci.org/core-roadmap-towards-v10/#rfc-32-projects). Being able to automate team configuration is a pretty clear next step.

The basic thought is that we can go beyond "configuration as code" for pipelines and towards "configuration as code" for the entire Concourse cluster. Pushing all of this into Concourse means *pipelines* become the API, and instead of having bespoke handcrafted scripts run `fly` commands on developer machines, you're building on existing Concourse abstractions in order to automate your own entire stack.

This idea is basically doubling down on the idea of Concourse as a workflow automation tool, so that it can be used as the source of truth and policy enforcement from the top-down rather than having to maintain Concourse *plus* your own bespoke automation.

> * We automate the monitoring of specific build steps/jobs/tasks/etc for compliance reasons, using an admin local user (have to login to each team individually, **over 150 teams**).
>   * For each team, we use `get-pipeline` with `--json` on each pipeline, and run our scans (**2,700 pipelines** as of a week or so ago).

This sounds like an interesting use case to explore - I would argue we can do better than bulk-scanning all pipelines on an interval. :thinking: Gonna leave this one up in the air though since it's a bit of a side-tangent.

> * For each team and unpaused pipeline, we also check build logs to see when the last run was. If it was a very long time ago, we pause the pipeline (reducing resource checks, given we have so many pipelines).

This would be addressed by #4673 once we implement [pipeline archiving](https://project.concourse-ci.org/projects/MDc6UHJvamVjdDM4Mzc3ODI=).

> * We have product teams (product team maps to a concourse team, in this scenario), create/destroy pipelines for specific branches and whatnot (and the new things in 5.8.0 could potentially help with this), in addition to their own automation from external systems for new apps/apis/etc. We don't want to give these users **keys to the kingdom**.

If your policy for pipeline automation were itself written in a pipeline using the `set_pipeline` step, you wouldn't have to worry about handing keys to anyone at all.

> * We have external systems (not managed by the team that manages Concourse) who create pipelines based off order forms/automated event systems, based off the team name (i.e. ability to give them access to create pipelines for specific teams without giving them access to all, e.g. one very sensitive team, is difficult). This type of team isn't a product team, more a group that manages pipelines for other teams.

This also sounds like something that could be automated building on some of the ideas I already mentioned (`set_pipeline` step, perhaps `set_team` step, etc. - downstream of a resource or perhaps using an [`across` step](https://github.com/concourse/rfcs/pull/29) to configure the full set of desired teams/pipelines in one go).

---

My goal here isn't to deflect this issue off to other much larger issues - it would still be nice to be able to log in without a browser. My goal here is only to point out that once you start talking about the real needs at hand, there are often many other ways to approach them, and in our case we often already have ideas 'in the works' that may address those needs more completely.

> tl;dr: things like forcing browser access, having undocumented API's, etc, seems (IMO) as though it's cutting off a good amount of the potential for Concourse, primarily in the enterprise space. I feel as though adoption would be much greater if there was a larger, more beginner-friendly ecosystem, as more companies move towards more automation. Not all Concourse users are going to be the stereotypical "using a macbook at a coffee shop, sipping my latte, writing my nodejs with emojis in each git commit message "

The challenge is this: by opening up the API, we would have more surface area to support in *addition* to the pipelines API, and the existence of that API delays the discovery of use cases such as your own. IMO this stunts the growth of Concourse as a workflow automation tool. Concourse is supposed to *be* the automation tool, not be what you *automate*.

By digging deep and figuring out what exactly y'all are trying to do, I hope to integrate these use cases into Concourse itself so that we don't have 20 different enterprise companies implementing 20 different solutions for automating Concourse with varying degrees of stability and testing.

...Sorry for my counter-rant. :sweat_smile:

**Comment by krishnam-pakalapati-git** [comment]
My stab on this issue:

**Approach #1:**
Current Status: Yet to try
Given Concourse works on the concept of resources (like plugins), It would be most suitable approach to build a new resource that can be integrated into Concourse and extend support for Octane -  https://concourse-ci.org/implementing-resource-types.html 

Pros:
Seamless integration to Concourse.
No Backward compatibility issues when Concourse upgrades are considered.

Cons:
This would require investing decent amount of time.
Need high end resources and Dev support.
Skill sets required to build resource – Java / GoLang

It may also be looked at as a contribution back to Concourse community to deal with similar problems/usecases if we build this resource.

**Approach #2** – Define a Concourse Job to run our desired automation and publish results back 
Current Status: Yet to try
We need to create Concourse jobs that does the below things.
-	Jobs that runs your tests after build-deploy phases are done. This job will also publish the results to Octane using Octane API - https://admhelp.microfocus.com/octane/en/15.0.20/Online/Content/API/articles_API2.htm
-	When we need to retrieve the results for a specific test run from Octane, you can create a Concourse job that fetches them and analyzes them.
We can accomplish this most simply via Concourse tasks (https://concourse-ci.org/tasks.html)
ALM Octane Concourse resource (https://concourse-ci.org/resource-types.html , https://concourse-ci.org/implementing-resource-types.html)

Pros:
Seamless integration to Concourse.
No Backward compatibility issues when Concourse upgrades are considered.

Cons:
One way solution. Concourse  Octane
Does not support Triggering jobs from Octane

Approach #3:
Current Status: BLOCKED – Unable to generate concourse auth token in an automated way.
Build a middleware using
 Trigger REST API
 Automated fly commands

2.1 REST API:  Middleware SDK that we build talks to API of both Octane and Concourse.
Pro's:
Relatively easy to attempt than building a new Concourse Resource
Cons:
Concourse API is not documented or supported - this means any automation we make could break when we upgrade concourse. We could think of set_pipeline step and the resource interface. I also think that we don't have to deploy another cloud tool, but can instead have concourse (which you have already deployed) do what you need, which may make our life easier - https://concourse-ci.org/jobs.html#schema.step.set-pipeline-step.set_pipeline.

2.2 Automate fly commands: Middleware interacts with fly console and does the end-to-end orchestration of reading jobs, pipelines, triggering jobs, publishing results etc. 
Pro's:
Relatively easy to attempt than building a new Concourse Resource
Cons: 
Cannot generated concourse auth token due to limitation that we have been discussing in length.

To deal with this;  as an alternative (not so seamless and clean approach as I already admitted), we can build a Headless Broweser script that takes care of below 4 steps. 

1.	Trigger fly login with option --open-browser
2.	Do SSO login via git authentication. This generates concourse auth token
3.	Copy Bearer Token clicking on copy token to clipboard button
4.	Feed this token to fly prompt: Enter Bearer token

Now, do end-to-end orchestration of reading jobs, pipelines, triggering jobs, publishing results etc using above saved token by invoking **fly** commands.
 
What is the issue with this alternative approach?

**Comment by vito** [comment]
Quick update on this: https://github.com/concourse/concourse/issues/5487 was opened which sounds like another example where a `set_team` step would help. 🤔

**Comment by krishnam-git** [comment]
Can someone shed light on how _set_pipeline_ could help with an automated login from a third-party tool of script?

As far as I know, there are a bunch of oauth flows that were never implemented support for. Once the generic tokens work is done, we'll be able to use a GitHub token (when login via GitHub) automatically to get a useful concourse token that can be used for automated login.

**Comment by vito** [comment]
> Can someone shed light on how _set_pipeline_ could help with an automated login from a third-party tool of script?

The [`set_pipeline` step](https://concourse-ci.org/jobs.html#set-pipeline-step) doesn't require auth, you just run it in a build and it configures the pipeline within the same team as the build. So if one of your reasons for requiring automated logins is to script `fly set-pipeline` or use the hopefully-eventually-deprecated `concourse-pipelines` resource, you could explore using the step instead.

> As far as I know, there are a bunch of oauth flows that were never implemented support for. Once the generic tokens work is done, we'll be able to use a GitHub token (when login via GitHub) automatically to get a useful concourse token that can be used for automated login.

I don't think this is true, but @pivotal-jwinters can maybe clear that up. As far as I know the generic tokens work just makes it so we use the tokens issued by [Dex](https://github.com/dexidp/dex), our embedded auth component, rather than minting our own.

**Comment by ari-becker** [comment]
Just to leave a comment about potholes in prior art that I've discovered while working on other stuff -

[saml2aws](https://github.com/Versent/saml2aws) is a project (like `fly`) written in Go, that needs to authenticate with external OIDC providers. Unlike `fly`, it handles the authentication process entirely within the terminal, and tries not to enter a browser.

Still, when authenticating with G Suite, it prompts the user to open a link in a browser to view a Captcha and type it in via the command line.

Additionally, I'm currently stuck with that tool since we use Yubikeys on our G Suite accounts, and the tool tries to handle U2F from the command line. To do so, it uses the [go-u2fhost](https://github.com/marshallbrekka/go-u2fhost) library, which currently works on OS X but is broken on Linux due to arcane issues with how hid and libusb identify U2F security keys.

The impression I'm getting from this is that the engineering for implementing login flows dependably across identity providers with the range of login methods which those identity providers permit is not yet mature (at least in Go, though I imagine the situation in other language ecosystems isn't much better). Using the browser is much more dependable in practice, at least at this point in time.

**Comment by krishnam-git** [comment]
> > Can someone shed light on how _set_pipeline_ could help with an automated login from a third-party tool of script?
> 
> The [`set_pipeline` step](https://concourse-ci.org/jobs.html#set-pipeline-step) doesn't require auth, you just run it in a build and it configures the pipeline within the same team as the build. So if one of your reasons for requiring automated logins is to script `fly set-pipeline` or use the hopefully-eventually-deprecated `concourse-pipelines` resource, you could explore using the step instead.
> 
> > As far as I know, there are a bunch of oauth flows that were never implemented support for. Once the generic tokens work is done, we'll be able to use a GitHub token (when login via GitHub) automatically to get a useful concourse token that can be used for automated login.
> 
> I don't think this is true, but @pivotal-jwinters can maybe clear that up. As far as I know the generic tokens work just makes it so we use the tokens issued by [Dex](https://github.com/dexidp/dex), our embedded auth component, rather than minting our own.

We could simply create a task that could deal with the external system interaction within Concourse itself and obviously, we dont need to deal with concourse login this way. Well, that's not going to unblock automated login to concourse from external systems.

Consider these simple use cases which needs automated login externally.
Lets say, QualityCenter or ALM wants to trigger some pipelines from their own eco system and wait for job results to get published back. 
An external tool that can automate creating and updating multiple pipelines in a batch for a newly onboarded team.

**Comment by tegataiprime** [comment]
For the moment, we're using Selenium to automate a service account login via browser & LDAP.  We grab the bearer token and then perform a fly login.  

However, we have to update our login script whenever the format or name of the bearer token changes.  We would really appreciate a more sustainable solution for a headless login via fly.

**Comment by speculatrix** [comment]
also see https://github.com/concourse/concourse/issues/5726
where I explain in detail why I want to be able to do command line login with ldap authentication

**Comment by vito** [comment]
@speculatrix Browserless login is one of those features where it's a seemingly simple surface-level change that unlocks all sorts of use cases. However the technical and support implications are far greater than the change implies - as evidenced by this gigantic thread - and we're the ones on the hook for them. (P.S. I don't think you were outright trying to be condescending, but I'd like to ask you to be more conscious of your tone - there's a reason we ask these questions. 🙂)

Re: the web interface, that sounds like really useful information that we should add to the product somehow rather than forcing everyone to come up with their own bespoke monitoring stack.

Re: monitoring, is the existing monitoring infrastructure insufficient? (Presumably not - why not?)

Re: automated pipeline deployment, I'm not quite sure what you mean by this, but it kind of sounds like there's some overlap with OPA (https://github.com/concourse/rfcs/pull/41 + https://github.com/concourse/concourse/pull/5034) and perhaps more broadly the [v10 roadmap](https://blog.concourse-ci.org/core-roadmap-towards-v10/).

In short, browserless login is likely a significant technical investment - we use [Dex](https://github.com/dexidp/dex) which doesn't seem to have a clear path forward for this. Following that investment, there is also a continued tax to pay through API support, which we'd like to avoid so that we can focus on the "real" API, which is pipelines/YAML/etc.

Whenever we ask what people are trying to do it tends to reveal that many people are trying to solve the same problems in different ways, or many people are trying to work around deficiencies elsewhere in Concourse. We would prefer to address the the core of these use cases so that we can minimize the amount of cruft people have to build around Concourse and minimize our own support burden.

**Comment by vito** [comment]
But as a quick follow-up, if someone were to graciously spend their time and LDAP experience to see there's a way for Dex to somehow support a browser-less login flow after all, I'd happily merge a PR for it, since it seems totally reasonable to expect `fly login -u -p` to work for LDAP. My reply above more broadly applies to the difficulty of supporting browserless login for *all* connectors. I still believe it holds true for LDAP, but I think LDAP is a more clear-cut case of "this should work."

**Comment by speculatrix** [comment]
What is Dex?

**Comment by speculatrix** [comment]
You asked: "is the existing monitoring infrastructure insufficient?"
Are you referring to this? https://concourse-ci.org/metrics.html

btw, on that page "the metrics from our deployment are public" is broken, so I can't see what your metrics look like.

we run a self-hosted zabbix instance which is cheap and very effective, and it's easy to add plugins.  Just 76 lines of bash, maybe 50 if I squashed it down, allows me to measure and graph pipelines, workers and containers in their different states, I simply run "fly" and parse the output to get the numbers. It might not be the most efficient way but it works.

**Comment by vito** [comment]
@speculatrix Yeah - looks like we need to update that link - we use DataDog now. Our dashboard can be found [here](https://p.datadoghq.com/sb/bb663c3fb-de49345af28e9fab7b850b0201eb2e9b).

I haven't heard of Zabbix, but Concourse supports emitting to Prometheus, DataDog, InfluxDB, NewRelic, and as a fallback - directly to logs. It looks like there's some sort of [Prometheus Zabbix integration](https://www.zabbix.com/documentation/current/manual/config/items/itemtypes/prometheus) which may be helpful.

[Dex](https://github.com/dexidp/dex) is an open source identity provider that we bake in to Concourse to support logging in to any connector that Dex supports without having to reinvent the wheel and implement all of it ourselves. But it's a double-edge sword; our login flow is limited to whatever Dex supports - and I'm not sure if it has a mechanism for logging in without a browser.

**Comment by speculatrix** [comment]
for those who stumble on this thread looking for concourse and zabbix:
https://github.com/speculatrix/zabbix-concourse-ci

**Comment by DonBower** [comment]
our security team has asked us to remove "local" users.
Therefore we desperately need to be able to provide an LDAP userid/password to log into the fly cli.  Frankly, i was shocked to see this does not work.

**Comment by vito** [comment]
As I mentioned in https://github.com/concourse/concourse/issues/3208#issuecomment-641348909 I would totally accept a PR that adds support for `fly login -u -p` for LDAP auth. I'm not sure if [Dex](https://github.com/dexidp/dex) supports it, so it might be best to ask there first.

The sad truth is if we were to work on this, it would mean de-prioritizing the v10 roadmap. A lot of people are shocked that Concourse doesn't support branches and pull requests. This functionality is more important to a larger amount of people and has already dragged on for way too long. We have to choose our battles and enable the community to fill the gaps where we can't.

I also want to be point out that the things we're working on *are* connected to this in some way, and it's not our intent to put on blinders and totally ignore the problem - for example, the Projects RFC now has an open question about [using Projects to replace Teams](https://github.com/vito/rfcs/blob/5679ee86b4e810a8ae8e00dcb21aca2ad745103e/032-projects/proposal.md#can-projects-replace-teams), which would effectively provide full-stack GitOps automation from within Concourse itself, eliminating a lot of the need for automated `fly` login while meeting a whole lot of other use cases and minimizing our burden in terms of API support.

At this point I'm starting to feel like a bit of a broken record, probably because there's too much text to read at this point before folks reply. I don't blame anyone for being "shocked" that we don't support this, since I know it can be frustrating that seemingly simple functionality just does not work, but again - there are many other parts of Concourse that need our focus, and we have to choose our battles and hope that others can contribute so that the weight of an entire industry doesn't fall on the shoulders of our development team.

I'm going to close this because it really just doesn't feel productive to leave open anymore. I'm also going to lock this so that this comment doesn't get lost in a sea of bump comments, as this issue can be a bit of a Google honeypot. There is a ton of nuance to this discussion, but at this point it feels like everything that needs to be said has already been said, and there's really not much else to talk about. I'm sorry, but we just can't justify prioritizing this.

If anyone wants to contribute their time and energy to improving this we're more than happy to help in a Discussion or PR.

**Comment by vito** [comment]
I spent a bit more time looking into this to provide guidance to anyone who wanted to help.

In a darkly hilarious twist of fate, it looks like supporting browserless LDAP auth is just a tiny change to make this value configurable:

https://github.com/concourse/concourse/blob/c4261d94405c3a7cc200b7ed8bd3f0ed84b12b4b/skymarshal/dexserver/dexserver.go#L106

I was able to change this const value to `"ldap"`, re-build Concourse, point it to a [public LDAP test server](https://www.forumsys.com/tutorials/integration-how-to/ldap/online-ldap-test-server/), and log in with `fly login -u -p`.

Funnily enough, we're the ones that [submitted the PR to Dex](https://github.com/dexidp/dex/pull/1621) to support this in the first place. Granted, it was based on someone else's abandoned PR, which might explain why we didn't connect the dots to see that this same change applies to other connectors beyond the local auth use case and that the value should clearly be configurable. That being said, only 3 connectors support it: LDAP, Atlassian Crowd, and Keystone, so this isn't a general solution to the "browserless login" problem.

Anyhow, I don't have time to tidy this up and submit a PR myself, but it should be easy enough if someone wants to pick it up!
